plugins/scaffolder-backend/src/scaffolder/stages/publish/helpers.ts

```diff
@@ -67,3 +68,63 @@ export async function initRepoAndPush({
     remote: 'origin',
   });
 }
+
+type BranchProtectionOptions = {
+  client: Octokit;
+  owner: string;
+  repoName: string;
+  isRetry?: boolean;
+};
+
+export const enableBranchProtectionOnDefaultRepoBranch = async ({
+  repoName,
+  client,
+  owner,
+  isRetry = false,
+}: BranchProtectionOptions): Promise<
+  RestEndpointMethodTypes['repos']['updateBranchProtection']['response']
+> => {
+  const { data: repo } = await client.repos.get({
+    owner,
+    repo: repoName,
+  });
+
+  try {
+    const response = await client.repos.updateBranchProtection({
+      headers: {
+        Accept:
+          /**
+           * 👇 we need this header because allowing a custom
+           * reviewer count on branch protection is a preview
+           * feature.
+           *
+           * More here: https://docs.github.com/en/rest/overview/api-previews#require-multiple-approving-reviews
+           */
+          'application/vnd.github.luke-cage-preview+json',
+      },
+      owner,
+      repo: repoName,
+      branch: repo.default_branch,
```

Comment:
Omitting this should use the default branch right? So we can skip the extra API call above. GitHub rate limits are a precious commodity 😅

Reply:
This isn't something we can omit, unfortunately. See [GitHub's API docs](https://octokit.github.io/rest.js/v18#repos-update-branch-protection) for more info. Additionally, the `default_branch` for a new repo is `main` by default, but the scaffolder pushes a branch called `master` instead after `GithubPublisher.publish`: 

https://github.com/backstage/backstage/blob/99dddfb61001ee823e717047fc3e9e9954a86b04/plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/github.ts#L216

So the order of requests is as follows:
1. [Create a repo](https://github.com/backstage/backstage/blob/99dddfb61001ee823e717047fc3e9e9954a86b04/plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts#L118-L133). There is no default branch yet.
2. [Push code to the `master` branch](https://github.com/backstage/backstage/blob/99dddfb61001ee823e717047fc3e9e9954a86b04/plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts#L88-L96), setting it as the default.
3. [Get the latest state of the repo and infer the default branch](https://github.com/TejasQ/backstage/blob/86e63c23c6ed85fdb78b15e7617edf2db7adc176/plugins/scaffolder-backend/src/scaffolder/stages/publish/helpers.ts#L87-L90).

There's definitely a way to optimize this by `const`antizing the name of a scaffolded default branch and passing a reference to it throughout the codebase. Let me know if you'd prefer that.

Reply:
Oh, so the branch is already hardcoded to always be `master`? Let's just use that then :grin:
Yeah could be good to put it in a constant so that future changes to the way branches are handled don't miss any spot.

Reply:
I feel this optimization _might_ be outside of the scope of this PR though, since there are 315 occurrences of `master` in the code of this repo. Would it be okay to do that after we merge this PR?

Reply:
We'd only only do it for the GitHub publisher though. The publishers are somewhat decoupled from each other and we wouldn't want to enforce the same default branch across the entire project.

Let's get it done in this PR since it's the one wanting to add the unnecessary call. If you want you could just skip the constant and put a literal there if it's causing trouble.

Reply:
Got it, thanks a lot! 🙌

Reply:
I've used a literal instead because `'master'` as a literal is used pretty much everywhere – even in YAML files – which signals to me that making it a dynamic value might be the basis of a future discussion. :D